File: go/spiral-matrix/spiral_matrix.go
package spiralmatrix

func SpiralMatrix(size int) [][]int {
    if size < 1 {
        return [][]int{}
    }

    x := 0
    y := 0
    dir := "right"
    spiral := [][]int{}

    for i := 0; i < size; i++ {
        row := []int{}
        for r := 0; r < size; r++ {
            row = append(row, 0)
        }
        spiral = append(spiral, row)
    }

    for i := 0; i < size * size; i++ {
    	spiral[y][x] = i + 1

    	switch dir {
    	case "right":
        	x++
            if x >= size || spiral[y][x] != 0 {
                x--
                dir = "down"
                y++
        	}

        case "left":
            x--
            if x < 0 || spiral[y][x] != 0 {
                x++
                dir = "up"
                y--
            }
            
        case "up":
            y--
            if y < 0 || spiral[y][x] != 0 {
                y++
                dir = "right"
                x++
            }

      	case "down":
        	y++
        	if y >= size || spiral[y][x] != 0 {
          		y--
          		dir = "left"
          		x--
        	}
    	}
  }

    return spiral
}

File: javascript/spiral-matrix/spiral-matrix.js
export const spiralMatrix = (size) => {
  if (size < 1) {
    return []
  }

  let x = 0
  let y = 0
  let direction = 'right'
  const spiral = []

  for (let i = 0; i < size; i++) {
    let row = []

    for (let j = 0; j < size; j++) {
      row.push(null)
    }

    spiral.push(row)
  }

  for (let i = 0; i < size * size; i++) {
    spiral[y][x] = i + 1

    switch(direction) {
      case 'right':
        x++
        if (x >= size || spiral[y][x] !== null) {
          x--
          direction = 'down'
          y++
        }
        break

      case 'left':
        x--
        if (x < 0 || spiral[y][x] !== null) {
          x++
          direction = 'up'
          y--
        }
        break

      case 'up':
        y--
        if (y < 0 || spiral[y][x] !== null) {
          y++
          direction = 'right'
          x++
        }
        break
      case 'down':
        y++
        if (y >= size || spiral[y][x] !== null) {
          y--
          direction = 'left'
          x--
        }
        break
    }
  }

  return spiral
}